src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
 
+    if (action.type === 'clear') {
+      currentState = {};
+    }
```

Comment:
You should use a switch statement instead of multiple if statements when you have a limited amount of conditions to check. This will make the code cleaner and more maintainable. Additionally, a default case should be added for error handling. [CODE STYLE]